### PR TITLE
variable scopes 

### PR DESCRIPTION
**Pull request by justinbmeyer**
for #181

### CODE REVIEW

**Comment by matthewp** [comment]
Is this to support a let helper in stache?

File: can-view-scope.js [hunk 1]
```diff
@@ -34,30 +44,155 @@ function Scope(context, parent, meta) {
...
+			return false;
+		};
+	},
+	makeShouldSkipSpecialContexts: function(parentContextWalkCount){
+		var walkCount = parentContextWalkCount || 0;
+		return function(currentScope){
+			if(currentScope.isSpecial()) {
+				return true;
+			}
+			walkCount--;
```
Comment (phillipskevin) [inline can-view-scope.js:204]: Can you explain why this is decrementing the `walkCount`? I'm not sure exactly what's going on here.

File: can-view-scope.js [hunk 2]
```diff
@@ -219,99 +359,94 @@ assign(Scope.prototype, {
...
+						updateSetObservable = false;
+					if(isVariableScope === true && nameIndex === 0) {
+						// we MUST have pre-defined the key in a variable scope
+						updateSetObservable = canReflect.hasOwnKey( parentValue, keyReads[nameIndex].key);
+					} else {
+						updateSetObservable =
+							// Has more matches
+							nameIndex > setObserveDepth ||
+							// The same number of matches but it has the key
+							nameIndex === setObserveDepth && (typeof parentValue === "object" && canReflect.hasOwnKey( parentValue, keyReads[nameIndex].key));
```
Comment (phillipskevin) [inline can-view-scope.js:389]: Will using `hasOnwKey` cause an issue with functions? I vaguely remember `keyReads[nameIndex].key in parentValue` being used in order to support functions.

File: can-view-scope.js (hunk 1 above)
Comment (justinbmeyer) [reply can-view-scope.js:204]: This is used for something like `../../../foo`, which would result in a `parentContextWalkCount` of 3, one for each `../`.  

It uses that number as a counter to know when it should start searching for `foo`.

File: can-view-scope.js (hunk 2 above)
Comment (justinbmeyer) [reply can-view-scope.js:389]: I'm not sure what you mean.  Can you explain more?  What kinds of issues?  Do you mean this should be `canReflect.hasKey()`? Yeah, might need to be `hasKey` to support values on the proto chain (which is what I think you mean by "functions").

Comment (phillipskevin) [reply can-view-scope.js:389]: I just mean functions on the context so things like `{{ func() }}` will work.

I ask becuase there's this difference between using `in` and using `hasKey`:
```js
const Foo = DefineMap.extend({
  func() {}
});

const foo = new Foo();

Reflect.hasOwnKey(foo, "func"); // false
Reflect.hasKey(foo, "func"); // false
"func" in foo; // true
```

Maybe this is should be fixed in DefineMap, but that is how it works currently.

Comment (phillipskevin) [reply can-view-scope.js:389]: so is this not going to be a problem @justinbmeyer ?

Comment (justinbmeyer) [reply can-view-scope.js:389]: I thought I changed it to hasKey.  It probably will be a problem.

Comment (justinbmeyer) [reply can-view-scope.js:389]: Changed it here: https://github.com/canjs/can-view-scope/commit/02a0c6c1a5a5f4a454241b97b83447618fc40307#commitcomment-30467125

**Pull request merged**